modules/stampcollection.py

```diff
@@ -78,16 +82,54 @@ def calculate_stamps(self):
             toi = self.utils.index[to_id]
             total_votes_by_user = self.utils.get_votes_by_user(from_id)
             if total_votes_by_user != 0:
-                score = (self.gamma * votes_for_user) / total_votes_by_user
+                score = votes_for_user / total_votes_by_user
                 users_matrix[toi, from_id_index] = score
-        for i in range(1, user_count):
-            users_matrix[i, i] = -1.0
-        users_matrix[0, 0] = 1.0
+        for i in range(user_count):
+            users_matrix[i, i] = 0
+
+        # self.log.debug(self.class_name, matrix=users_matrix)
 
         user_count_matrix = np.zeros(user_count)
         user_count_matrix[0] = 1.0  # God has 1 karma
 
-        self.utils.scores = list(np.linalg.solve(users_matrix, user_count_matrix))
+        scores = user_count_matrix
+        drains = self.utils.get_total_drains()
+        decay = 1 - self.DECAY
+        # self.log.debug(self.class_name, msg="There is" + (" not" if not drains else "") + " a drain!")
+        for i in range(self.MAX_ROUNDS):
+            old_scores = scores
+            scores = np.dot(users_matrix, scores) * decay
+            if drains:  # If there are drains, we need to make sure stampy always has 1 trust.
+                scores[0] = 1
+            # self.log.debug(self.class_name, step=scores)
+
+            # Check if solved
+            solved = True
+            for a, b in zip(old_scores, scores):
+                if not round(a, self.PRECISION) == round(b, self.PRECISION):
+                    solved = False
+                    break
+            if solved:
+                # Double check work.
+                solved = False
+                for a in scores:
+                    if round(a, self.PRECISION) != 0:
+                        solved = True
+                        break
```

Comment:
what's the meaning why all scores must be non-zero, please? (alternative question, why is it bad if at least 1 score is 0?)

Reply:
This actually checks to make sure not everything is 0.  If the karma goes in a loop it just decays forever and the entire system resolves to 0, which is not correct.  If one person is 0 that is fine.  So we could probably simplify this with the np.all thing with the other option being 0.

Reply:
ah, would `a.round(self.PRECISION).count_nonzero()` do the same?

Reply:
No, that would result in `AttributeError: 'numpy.ndarray' object has no attribute 'count_nonzero'` but `np.count_nonzero(a.round(self.PRECISION)) > 0` would :joy:  I already wrote a patch doing `solved = np.any(scores.round(self.PRECISION) != 0)`